README
================================================================================
ruby-pg: Ruby interface to PostgreSQL RDBMS
================================================================================

This library is copyrighted by the authors.

Authors:
  * Yukihiro Matsumoto <matz@ruby-lang.org>
      Author of Ruby.
  * Eiji Matsumoto <usagi@ruby.club.or.jp>
      One of users who loves Ruby.
  * Jeff Davis <ruby-pg@j-davis.com>

Thanks to:
  * Noboru Saitou <noborus@netlab.jp>
      Past maintainer.
  * Dave Lee
      Past maintainer.
  * Guy Decoux   ts <decoux@moulon.inra.fr> 
  
Maintainer:
  Jeff Davis <ruby-pg@j-davis.com>

Copying:
  You may redistribute this software under the terms of the GPL,
  included in the file named GPL; or under the same terms as Ruby,
  included in the file name LICENSE.

- Summary

This is the extension library to access a PostgreSQL database from Ruby.
This library works with PostgreSQL 7.4 and later.

- Requirements

  Ruby 1.8 or later.
  PostgreSQL 7.4 or later installed.

  It may work with earlier versions as well, but those are
  not regularly tested.

- How to install ?

Follow the instructions below to compile and install:

  ruby extconf.rb
  make
  su              (if necessary)
  make install

You may need to specify the directory name for the include files and the
-lpq library by using

    --with-pgsql-include=<include file directory>
    --with-pgsql-lib=<library directory>

For example:

  ruby extconf.rb --with-pgsql-include=/usr/local/pgsql/include \
                  --with-pgsql-lib=/usr/local/pgsql/lib

- Modules

'pg': The 'pg' module is the newer module, that has been greatly improved, and
  is almost a complete rewrite. It is not backwards compatible. Use this module
  for newly written code. It should be more stable, less buggy, and has more
  features.

'postgres': Older module, maintained for backwards compatibility. It 
  has known flaws that aren't reasonably fixable without breaking backwards
  compatibility. Use this module if you have code that already works, and 
  you just want the fixes that I've committed to this module (for instance,
  this module is compatible with PostgreSQL 8.3).

- How to use ?

This gem builds and installs two PostgreSQL database adapters, 'postgres'
and 'pg'. 'postgres' is currently (Jan 2008) the recommended PostgreSQL
adapter for use with Rails.

The standard way to download and install the most current stable  
version of the postgres gem (from http://gems.rubyforge.org) is to use  
the RubyGem package manager. You may need to supply RubyGem with the 
location of the libpq library and the libpq.h and libpq/libpq-fs.h 
files, and may need to run as root.
If you installed from source on a Unix system you can locate these  
libpq files with:
find <path to install directory> -name "libpq-fe.h" -print
With binary distributions, you may need to install additional  
PostgreSQL development libraries to get these files.

Then run:
sudo gem install postgres -- --with-pgsql-include-dir=<location of  
Postgresql>/include --with-pgsql-lib-dir=<location of Postgresql/lib

Example:
on Mac OS X with PostgreSQL in /Library/PostgreSQL8 use
--with-pgsql-include-dir=/Library/PostgreSQL8/include --with-pgsql-lib- 
dir=/Library/PostgreSQL8/lib

To use the postgres adapter with Rails:
refer to it as
adapter: postgresql
in your database:yaml file

The pg adapter cannot be used with Rails yet and is available for  
testing and developer use.

To use these modules in Ruby directly (not Rails), refer to the RDoc
documentation.

- Acknowledgments

We are thankful to the people at the ruby-list and ruby-dev mailing lists.
And to the people who developed PostgreSQL.

- Copying

This library is copyrighted by its authors; Yukihiro Matsumoto, and Eiji
Matsumoto.

Portions copyright Laika, Inc.





ChangeLog
Ruby extension library to access PostgreSQL.
===========================================

Tue Dec 14 2004
	* set_notice_processor: add methods set_notice_processor.
	Dennis Vshivkov <walrus@amur.ru>	
	
Sat Nov  6 2004
	* unescape_bytea: add methods unescape_bytea.
	  patch supplied by "Henry T. So Jr." <henryso@panix.com>
	
Sat Oct  9 2004
	* extconf.rb: VERSION -> RUBY_VERSION. 
	  add header check.
	  patch supplied by MoonWolf <moonwolf@moonwolf.com>

Fri Jul  2 2004
	* open: add methods open.
	  patch supplied by Leon Brooks <leon-ruby-postgres@cyberknights.com.au>
	
Tue Feb 24 2004
	* postgres.c: add comments.
	  patch supplied by Gavin Kistner <gavin@refinery.com>
	
Fri Nov 23 2003
	* extconf.rb: add ssl support.
	  patch supplied by Martin Hedenfalk <mahe@kth.se>

Sat Nov  1 2003

	* oid: add methods oid.
	  patch supplied by Ceri Storey <cez@compsoc.man.ac.uk>
 
Mon Jan  6 2003 version 0.7.1
	
	* async_exec: remove check of PQisBusy.
	  patch supplied by <m_seki@mva.biglobe.ne.jp>
	
Wed Oct 16 2002 version 0.7.0
	
	* rename README.jp -> README.ja.

Thu Oct 10 2002
	
	* get_notify: bug fix Deletes of unnecessary "free(notify);"
	  debian bug report.
	
Wed Oct  9 2002
	
	* query: add pgresult_clear.
	  patch supplied by "Shirai,Kaoru" <shirai@p1jp.com>

Sat Sep 28 2002
	
	* PGconn: add methods escape, quote, escape_bytes. 
	  patch supplied by MoonWolf <moonwolf@moonwolf.com>

Tue Aug 20 2002 

	* extconf.rb add dir_config('postgres')
 	  patch supplied by Nate Haggard <nate@wordplace.com>
	
Mon Jun  3 2002 version 0.7.0-pre2
	
	* not to use fe_getauthname().
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>
	
Thu May 30 2002 version 0.7.0-pre1
	
	* Fix to insert_table so that nil values in Ruby are inserted
	  back into the database as nulls.
	  patch supplied by Jeremy Henty <jeremy@chaos.org.uk>

	* Fix to insert_table to protect characters that Postgres would
	  otherwise interpret specially.
	  patch supplied by Jeremy Henty <jeremy@chaos.org.uk>
	
Wed Apr 24 2002
	
	* Removed unused variables.
	* Improve notification code.
	* Add a missing "return Qnil" statement in a function 
	  declared to return VALUE.
	* Fix an minor error in pgconn_loopen().
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>	
	
Tue Mar  5 2002
	
	* fix pglarge_write return.
	  patch supplied by Noboru Matui <silicon@mx1.freemail.ne.jp>

Mon Jan  7 2002
	
	* refer to POSTGRES_INCLUDE and POSTGTRES_LIB 
	  environmental variables.
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>
	
Sun Dec 16 2001
	
	* add methods "getisnull".
	  patch supplied by Jeremy Henty <jeremy@chaos.org.uk>
	
Tue Nov 27 2001
	
	* convert NULL => nil .
	  patch supplied by "Shirai,Kaoru" <shirai@p1jp.com>

Mon Nov 19 2001 version 0.6.5 released
	
	* change my address.
	  Noboru Saitou <noborus@netlab.jp>
	
Fri Nov 16 2001	
	
	* default encoding set.
	  Riley <wormwood@speakeasy.org>

Fri Nov 16 2001	
	
	* add methods "cmdtuples"
          Gabriel Emerson <gemerson@evalsoft.com>	
	  Thanks to greentea@fa2.so-net.ne.jp
	
Thu Jul 12 2001
	
	* psql.rb $OPT_p => $OPT_p.to_i (bug fix).
	  Yuta TSUBOI <yuuta-t@is.aist-nara.ac.jp>
	
Mon Apr 23 2001 version 0.6.4 released
	
	* add include sys/types.h
          patch supplied by "Akinori MUSHA" <knu@iDaemons.org>.
	
Sun Apr 22 2001 version 0.6.3 released

	* document updated.
	
Sun Mar 18 2001
	
	* extern -> EXTERN changed.
          shibata <kshibata@vesta.ocn.ne.jp>
	
Sun Feb 11 2001
	
	* add methods "async_exec" and "async_query.
	  patch supplied by  Andy Yu <is@gnuchina.org>.

Tue Feb  6 2001 version 0.6.2 released
	
	* extconf.rb have_func second argument omit.
	  (old ruby version support).

Mon Feb  5 2001

	* RSTRING(*)->ptr -> STR2CSTR(*) changed.

Sun Feb  4 2001
	 
	* README and README.jp  copying are unified to Ruby's.
	 
Wed Jan 31 2001
	
	* postgres.html bug fix and better.
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>.

	* README bug fix and better.
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>.
	 
Sat Jan 27 2001
	
	* extconf.rb add error message for
	  Could not find PostgreSQL libraries.
	  patch supplied by Neil Conway <nconway@klamath.dyndns.org>.
	 
Tue Jan 23 2001

	* add methods "initialize".
	  patch supplied by <matz@zetabits.com>.

Sun Dec 17 2000 
	
	* 8192 byte limit removed.
	
Tue Oct 11 2000	

	* methods name lo* -> lo_* changed.

	* extconf.rb bug fix.
	   have_func("PQsetClientEncoding")

Thu Jun 30 2000 version 0.6.1 released

	* change receive parameter with VALUE.
	  (exec, insert_table, putline)

	* insert_table: remove debug print.
	
	* insert_table bug fix.
	  patch supplied by "Okada Jun" <yun@be-in.org>.

	* extconf.rb ClientEncoding->SetClientEnocoding.

Sun May 21 2000 version 0.6.0 released

	* change the maintainer with 
	  Noboru Saitou <noborus@mb.kcom.ne.jp>.

	* add provided clientEnocoding and setClientEncoding.

	* replace rb_str_new with rb_tainted_str_new.
	  ts <decoux@moulon.inra.fr>

	* add provided Large Object support!

	* error message changed PQerrorMessage() use.

Fri Feb 25 2000 version 0.5.4 released

	* PGError should be subclass of StandardError.

Fri Aug  6 1999 version 0.5.3 released

	* use PQsetdbLogin() instead of PQsetdb().
	  patch supplied by <kasa@air.linkclub.or.jp>.

Tue Jul 30 1999 version 0.5.2 released

	* document updated.
	* better Ruby 1.3 support.

Tue Jul 13 1999 version 0.5.1 released

	* modified for Ruby 1.3.x or later.

Wed Jan 20 1999 version 0.4.1 released

	* Makefile.rb, extconf.rb: trivial bug fix.

Thu Dec 12 1997 version 0.4 released

	* PGresult: add new method -- each{}, [] --
	* provided Makefile.rb for ruby-1.1b

Thu Sep 25 1997 version 0.3.1 released

	* extconf.rb bug fix.
	* documentation about 'How to install' 

Mon Sep 4 1997 version 0.3 released

	* remove restrictions about japanese EUC characters.
	* parseArgs related bug fix.

Mon Sep 2 1997 version 0.2 released


ext/compat.c
/************************************************

  compat.c -

  Author: matz 
  created at: Tue May 13 20:07:35 JST 1997

  Author: ematsu
  modified at: Wed Jan 20 16:41:51 1999

  $Author: jdavis $
  $Date: 2007-12-04 14:25:44 -0800 (Tue, 04 Dec 2007) $
************************************************/

#include <ctype.h>
#include "compat.h"

#ifdef PG_BEFORE_080300
int
PQconnectionNeedsPassword(PGconn *conn)
{
	rb_raise(rb_eStandardError, 
		"PQconnectionNeedsPassword not supported by this client version.");
}

int
PQconnectionUsedPassword(PGconn *conn)
{
	rb_raise(rb_eStandardError, 
		"PQconnectionUsedPassword not supported by this client version.");
}

int
lo_truncate(PGconn *conn, int fd, size_t len)
{
	rb_raise(rb_eStandardError, "lo_truncate not supported by this client version.");
}
#endif /* PG_BEFORE_080300 */

#ifdef PG_BEFORE_080200
int
PQisthreadsafe()
{
	return Qfalse;
}

int
PQnparams(const PGresult *res)
{
	rb_raise(rb_eStandardError, "PQnparams not supported by this client version.");
}

Oid
PQparamtype(const PGresult *res, int param_number)
{
	rb_raise(rb_eStandardError, "PQparamtype not supported by this client version.");
}

PGresult *
PQdescribePrepared(PGconn *conn, const char *stmtName)
{
	rb_raise(rb_eStandardError, "PQdescribePrepared not supported by this client version.");
}

PGresult *
PQdescribePortal(PGconn *conn, const char *portalName)
{
	rb_raise(rb_eStandardError, "PQdescribePortal not supported by this client version.");
}

int
PQsendDescribePrepared(PGconn *conn, const char *stmtName)
{
	rb_raise(rb_eStandardError, "PQsendDescribePrepared not supported by this client version.");
}

int
PQsendDescribePortal(PGconn *conn, const char *portalName)
{
	rb_raise(rb_eStandardError, "PQsendDescribePortal not supported by this client version.");
}

char *
PQencryptPassword(const char *passwd, const char *user)
{
	rb_raise(rb_eStandardError, "PQencryptPassword not supported by this client version.");
}
#endif /* PG_BEFORE_080200 */

#ifdef PG_BEFORE_080100
Oid
lo_create(PGconn *conn, Oid lobjId)
{
	rb_raise(rb_eStandardError, "lo_create not supported by this client version.");
}
#endif /* PG_BEFORE_080100 */

#ifdef PG_BEFORE_080000
PGresult *
PQprepare(PGconn *conn, const char *stmtName, const char *query,
	int nParams, const Oid *paramTypes)
{
	rb_raise(rb_eStandardError, "PQprepare not supported by this client version.");
}

int
PQsendPrepare(PGconn *conn, const char *stmtName, const char *query,
	int nParams, const Oid *paramTypes)
{
	rb_raise(rb_eStandardError, "PQsendPrepare not supported by this client version.");
}

int
PQserverVersion(const PGconn* conn)
{
	rb_raise(rb_eStandardError, "PQserverVersion not supported by this client version.");
}
#endif /* PG_BEFORE_080000 */

#ifdef PG_BEFORE_070400
PGresult *
PQexecParams(PGconn *conn, const char *command, int nParams, 
	const Oid *paramTypes, const char * const * paramValues, const int *paramLengths, 
	const int *paramFormats, int resultFormat)
{
	rb_raise(rb_eStandardError, "PQexecParams not supported by this client version.");
}

PGTransactionStatusType 
PQtransactionStatus(const PGconn *conn)
{
	rb_raise(rb_eStandardError, "PQtransactionStatus not supported by this client version.");
}

char *
PQparameterStatus(const PGconn *conn, const char *paramName)
{
	rb_raise(rb_eStandardError, "PQparameterStatus not supported by this client version.");
}

int
PQprotocolVersion(const PGconn *conn)
{
	rb_raise(rb_eStandardError, "PQprotocolVersion not supported by this client version.");
}

PGresult
*PQexecPrepared(PGconn *conn, const char *stmtName, int nParams, 
	const char * const *ParamValues, const int *paramLengths, const int *paramFormats,
	int resultFormat)
{
	rb_raise(rb_eStandardError, "PQexecPrepared not supported by this client version.");
}

int
PQsendQueryParams(PGconn *conn, const char *command, int nParams,
	const Oid *paramTypes, const char * const * paramValues, const int *paramLengths, 
	const int *paramFormats, int resultFormat)
{
	rb_raise(rb_eStandardError, "PQsendQueryParams not supported by this client version.");
}

int
PQsendQueryPrepared(PGconn *conn, const char *stmtName, int nParams, 
	const char * const *ParamValues, const int *paramLengths, const int *paramFormats,
	int resultFormat)
{
	rb_raise(rb_eStandardError, "PQsendQueryPrepared not supported by this client version.");
}

int
PQputCopyData(PGconn *conn, const char *buffer, int nbytes)
{
	rb_raise(rb_eStandardError, "PQputCopyData not supported by this client version.");
}

int
PQputCopyEnd(PGconn *conn, const char *errormsg)
{
	rb_raise(rb_eStandardError, "PQputCopyEnd not supported by this client version.");
}

int
PQgetCopyData(PGconn *conn, char **buffer, int async)
{
	rb_raise(rb_eStandardError, "PQgetCopyData not supported by this client version.");
}

PGVerbosity
PQsetErrorVerbosity(PGconn *conn, PGVerbosity verbosity)
{
	rb_raise(rb_eStandardError, "PQsetErrorVerbosity not supported by this client version.");
}

Oid
PQftable(const PGresult *res, int column_number)
{
	rb_raise(rb_eStandardError, "PQftable not supported by this client version.");
}

int
PQftablecol(const PGresult *res, int column_number)
{
	rb_raise(rb_eStandardError, "PQftablecol not supported by this client version.");
}

int
PQfformat(const PGresult *res, int column_number)
{
	rb_raise(rb_eStandardError, "PQfformat not supported by this client version.");
}
#endif /* PG_BEFORE_070400 */

#ifdef PG_BEFORE_070300
size_t
PQescapeStringConn(PGconn *conn, char *to, const char *from, 
	size_t length, int *error)
{
	return PQescapeString(to,from,length);
}

unsigned char *
PQescapeByteaConn(PGconn *conn, const unsigned char *from, 
	size_t from_length, size_t *to_length)
{
	return PQescapeBytea(from, from_length, to_length);
}
#endif /* PG_BEFORE_070300 */


/**************************************************************************

IF ANY CODE IS COPIED FROM POSTGRESQL, PLACE IT AFTER THIS COMMENT.

***************************************************************************

Portions of code after this point were copied from the PostgreSQL source
distribution, available at http://www.postgresql.org

***************************************************************************

Portions Copyright (c) 1996-2007, PostgreSQL Global Development Group

Portions Copyright (c) 1994, The Regents of the University of California

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose, without fee, and without a written agreement
is hereby granted, provided that the above copyright notice and this
paragraph and the following two paragraphs appear in all copies.

IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY FOR
DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING
LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS
DOCUMENTATION, EVEN IF THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
ON AN "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATIONS TO
PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.

**************************************************************************/

#ifndef HAVE_PQSETCLIENTENCODING
int
PQsetClientEncoding(PGconn *conn, const char *encoding)
{
	char        qbuf[128];
	static const char query[] = "set client_encoding to '%s'";
	PGresult   *res;
	int         status;

	if (!conn || conn->status != CONNECTION_OK)
		return -1;

	if (!encoding)
		return -1;

	/* check query buffer overflow */
	if (sizeof(qbuf) < (sizeof(query) + strlen(encoding)))
		return -1;

	/* ok, now send a query */
	sprintf(qbuf, query, encoding);
	res = PQexec(conn, qbuf);

	if (res == NULL)
	return -1;
	if (res->resultStatus != PGRES_COMMAND_OK)
	status = -1;
	else
	{
		/*
		 * In protocol 2 we have to assume the setting will stick, and adjust
		 * our state immediately.  In protocol 3 and up we can rely on the
		 * backend to report the parameter value, and we'll change state at
		 * that time.
		 */
		if (PG_PROTOCOL_MAJOR(conn->pversion) < 3)
			pqSaveParameterStatus(conn, "client_encoding", encoding);
		status = 0;             /* everything is ok */
	}
	PQclear(res);
	return status;
}
#endif /* HAVE_PQSETCLIENTENCODING */

#ifndef HAVE_PQESCAPESTRING
/*
 * Escaping arbitrary strings to get valid SQL literal strings.
 *
 * Replaces "\\" with "\\\\" and "'" with "''".
 *
 * length is the length of the source string.  (Note: if a terminating NUL
 * is encountered sooner, PQescapeString stops short of "length"; the behavior
 * is thus rather like strncpy.)
 *
 * For safety the buffer at "to" must be at least 2*length + 1 bytes long.
 * A terminating NUL character is added to the output string, whether the
 * input is NUL-terminated or not.
 *
 * Returns the actual length of the output (not counting the terminating NUL).
 */
size_t
PQescapeString(char *to, const char *from, size_t length)
{
	const char *source = from;
	char	   *target = to;
	size_t		remaining = length;

	while (remaining > 0 && *source != '\0')
	{
		switch (*source)
		{
			case '\\':
				*target++ = '\\';
				*target++ = '\\';
				break;

			case '\'':
				*target++ = '\'';
				*target++ = '\'';
				break;

			default:
				*target++ = *source;
				break;
		}
		source++;
		remaining--;
	}

	/* Write the terminating NUL character. */
	*target = '\0';

	return target - to;
}

/*
 *		PQescapeBytea	- converts from binary string to the
 *		minimal encoding necessary to include the string in an SQL
 *		INSERT statement with a bytea type column as the target.
 *
 *		The following transformations are applied
 *		'\0' == ASCII  0 == \\000
 *		'\'' == ASCII 39 == \'
 *		'\\' == ASCII 92 == \\\\
 *		anything < 0x20, or > 0x7e ---> \\ooo
 *										(where ooo is an octal expression)
 */
unsigned char *
PQescapeBytea(const unsigned char *bintext, size_t binlen, size_t *bytealen)
{
	const unsigned char *vp;
	unsigned char *rp;
	unsigned char *result;
	size_t		i;
	size_t		len;

	/*
	 * empty string has 1 char ('\0')
	 */
	len = 1;

	vp = bintext;
	for (i = binlen; i > 0; i--, vp++)
	{
		if (*vp < 0x20 || *vp > 0x7e)
			len += 5;			/* '5' is for '\\ooo' */
		else if (*vp == '\'')
			len += 2;
		else if (*vp == '\\')
			len += 4;
		else
			len++;
	}

	rp = result = (unsigned char *) malloc(len);
	if (rp == NULL)
		return NULL;

	vp = bintext;
	*bytealen = len;

	for (i = binlen; i > 0; i--, vp++)
	{
		if (*vp < 0x20 || *vp > 0x7e)
		{
			(void) sprintf(rp, "\\\\%03o", *vp);
			rp += 5;
		}
		else if (*vp == '\'')
		{
			rp[0] = '\\';
			rp[1] = '\'';
			rp += 2;
		}
		else if (*vp == '\\')
		{
			rp[0] = '\\';
			rp[1] = '\\';
			rp[2] = '\\';
			rp[3] = '\\';
			rp += 4;
		}
		else
			*rp++ = *vp;
	}
	*rp = '\0';

	return result;
}

#define ISFIRSTOCTDIGIT(CH) ((CH) >= '0' && (CH) <= '3')
#define ISOCTDIGIT(CH) ((CH) >= '0' && (CH) <= '7')
#define OCTVAL(CH) ((CH) - '0')

/*
 *		PQunescapeBytea - converts the null terminated string representation
 *		of a bytea, strtext, into binary, filling a buffer. It returns a
 *		pointer to the buffer (or NULL on error), and the size of the
 *		buffer in retbuflen. The pointer may subsequently be used as an
 *		argument to the function free(3). It is the reverse of PQescapeBytea.
 *
 *		The following transformations are made:
 *		\\	 == ASCII 92 == \
 *		\ooo == a byte whose value = ooo (ooo is an octal number)
 *		\x	 == x (x is any character not matched by the above transformations)
 */
unsigned char *
PQunescapeBytea(const unsigned char *strtext, size_t *retbuflen)
{
	size_t		strtextlen,
				buflen;
	unsigned char *buffer,
			   *tmpbuf;
	size_t		i,
				j;

	if (strtext == NULL)
		return NULL;

	strtextlen = strlen(strtext);

	/*
	 * Length of input is max length of output, but add one to avoid
	 * unportable malloc(0) if input is zero-length.
	 */
	buffer = (unsigned char *) malloc(strtextlen + 1);
	if (buffer == NULL)
		return NULL;

	for (i = j = 0; i < strtextlen;)
	{
		switch (strtext[i])
		{
			case '\\':
				i++;
				if (strtext[i] == '\\')
					buffer[j++] = strtext[i++];
				else
				{
					if ((ISFIRSTOCTDIGIT(strtext[i])) &&
						(ISOCTDIGIT(strtext[i + 1])) &&
						(ISOCTDIGIT(strtext[i + 2])))
					{
						int			byte;

						byte = OCTVAL(strtext[i++]);
						byte = (byte << 3) + OCTVAL(strtext[i++]);
						byte = (byte << 3) + OCTVAL(strtext[i++]);
						buffer[j++] = byte;
					}
				}

				/*
				 * Note: if we see '\' followed by something that isn't a
				 * recognized escape sequence, we loop around having done
				 * nothing except advance i.  Therefore the something will
				 * be emitted as ordinary data on the next cycle. Corner
				 * case: '\' at end of string will just be discarded.
				 */
				break;

			default:
				buffer[j++] = strtext[i++];
				break;
		}
	}
	buflen = j;					/* buflen is the length of the dequoted
								 * data */

	/* Shrink the buffer to be no larger than necessary */
	/* +1 avoids unportable behavior when buflen==0 */
	tmpbuf = realloc(buffer, buflen + 1);

	/* It would only be a very brain-dead realloc that could fail, but... */
	if (!tmpbuf)
	{
		free(buffer);
		return NULL;
	}

	*retbuflen = buflen;
	return tmpbuf;
}
#endif

